SPHINXsys/src/shared/materials/diffusion_reaction.h

```diff
@@ -163,93 +163,126 @@ namespace SPH
 	 */
-	template <class BaseMaterialType = BaseMaterial, int NUM_SPECIES = 1>
+	template <class BaseMaterialType = BaseMaterial, int NUM_REACTIVE_SPECIES = 0>
 	class DiffusionReaction : public BaseMaterialType
 	{
+	public:
+		static constexpr int NumReactiveSpecies = NUM_REACTIVE_SPECIES;
 	private:
 		UniquePtrKeepers<BaseDiffusion> diffusion_ptr_keeper_;
+		SharedPtrKeeper<BaseReactionModel<NUM_REACTIVE_SPECIES>> reaction_ptr_keeper_;
 
 	protected:
-		typedef std::array<std::string, NUM_SPECIES> SpeciesNames;
-		SpeciesNames species_name_list_;
-		size_t number_of_species_;
-		std::map<std::string, size_t> species_indexes_map_;
-		StdVec<BaseDiffusion *> species_diffusion_;
-		BaseReactionModel<NUM_SPECIES> *species_reaction_;
+		typedef std::array<std::string, NUM_REACTIVE_SPECIES> ReactiveSpeciesNames;
+		StdVec<std::string> all_species_names_;
+		BaseReactionModel<NUM_REACTIVE_SPECIES> &reaction_model_;
+		std::map<std::string, size_t> all_species_indexes_map_;
+		StdVec<BaseDiffusion *> all_diffusions_;
+		IndexVector reactive_species_indexes_;
+		IndexVector diffusion_species_indexes_;
+		IndexVector gradient_species_indexes_;
 
 	public:
-		/** Constructor for material with diffusion only. */
-		template <typename... ConstructorArgs>
-		DiffusionReaction(SpeciesNames species_name_list, ConstructorArgs &&...args)
-			: BaseMaterialType(std::forward<ConstructorArgs>(args)...),
-			  species_name_list_(species_name_list),
-			  number_of_species_(species_name_list.size()),
-			  species_reaction_(nullptr)
+		/** Constructor for material with diffusion and reaction. */
+		template <typename... MaterialArgs>
+		DiffusionReaction(const StdVec<std::string> &all_species_names,
+						  SharedPtr<BaseReactionModel<NUM_REACTIVE_SPECIES>> reaction_model_ptr,
+						  MaterialArgs &&...material_args)
+			: BaseMaterialType(std::forward<MaterialArgs>(material_args)...),
+			  all_species_names_(all_species_names),
+			  reaction_model_(reaction_ptr_keeper_.assignRef(reaction_model_ptr))
 		{
-			BaseMaterialType::material_type_name_ = "Diffusion";
-			for (size_t i = 0; i != number_of_species_; ++i)
+			BaseMaterialType::material_type_name_ =
+				NUM_REACTIVE_SPECIES == 0 ? "Diffusion" : "DiffusionReaction";
+
+			for (size_t i = 0; i != all_species_names.size(); ++i)
 			{
-				species_indexes_map_.insert(make_pair(species_name_list[i], i));
+				all_species_indexes_map_.insert(make_pair(all_species_names[i], i));
+			}
+
+			for (size_t i = 0; i != NUM_REACTIVE_SPECIES; ++i)
+			{
+				const ReactiveSpeciesNames &reactive_species_names = reaction_model_.getSpeciesNames();
+				size_t reactive_species_index = all_species_indexes_map_[reactive_species_names[i]];
+				if (reactive_species_index != all_species_indexes_map_.size())
+				{
+					reactive_species_indexes_.push_back(reactive_species_index);
+				}
+				else
+				{
+					std::cout << "\n Error: reactive species '" << reactive_species_names[i] << "' not defined!" << std::endl;
```

Comment:
This logging and all other alike should be redirected to `cerr` instead of `cout`

Reply:
Ok. I will change it in the next revision.